bittide-instances/src/Bittide/Instances/Hitl/Driver/Demo.hs

```diff
@@ -0,0 +1,631 @@
+            liftIO $ putStrLn "Checking for MMIO access to MU CPUs over GDB..."
+            gdbExitCodes1 <- zipWithM muGdbCheck targets muGdbs
+            (gdbCount1, gdbExitCode1) <-
+              L.foldl foldExitCodes (pure (0, ExitSuccess)) gdbExitCodes1
+            liftIO
+              $ putStrLn
+                [i|MU GDB testing passed on #{gdbCount1} of #{L.length targets} targets|]
+            liftIO $ mapM_ ((errorToException =<<) . Gdb.loadBinary) muGdbs
+
+            liftIO $ mapM_ Gdb.continue ccGdbs
+            forM_ targets assertAllProgrammed
+            testResults <- awaitTestCompletions 60_000
+            (sCount, stabilityExitCode) <-
+              L.foldl foldExitCodes (pure (0, ExitSuccess)) testResults
+            liftIO
+              $ putStrLn
+                [i|Test case #{testName} stabilised on #{sCount} of #{L.length targets} targets|]
+
+            liftIO $ putStrLn "Getting UGNs for all targets"
+            ugnPairsTable <- zipWithM muGetUgns targets muGdbs
+            let
+              ugnPairsTableV = fromJust . V.fromList $ fromJust . V.fromList <$> ugnPairsTable
+            liftIO $ do
+              putStrLn "Calculating IGNs for all targets"
+              printAllIgns ugnPairsTableV fpgaSetup
+              mapM_ print ugnPairsTableV
+            currentTime <- liftIO $ muGetCurrentTime (L.head targets) (L.head muGdbs)
+            let
+              startOffset = currentTime + natToNum @(PeriodToCycles GthTx (Seconds StartDelay))
+              chainConfig :: Vec FpgaCount (Calc.PeConfig (Unsigned 64) (Index 9))
+              chainConfig =
+                Calc.fullChainConfiguration
+                  (SNat @3)
+                  fpgaSetup
+                  ugnPairsTableV
+                  startOffset
+            liftIO $ do
+              putStrLn [i|Current clock cycle: #{currentTime}|]
+              putStrLn "Calculated the following configs for the switch processing elements:"
+              forM_ chainConfig print
+            _ <- sequenceA $ L.zipWith3 muWriteCfg targets muGdbs (toList chainConfig)
+            liftIO $ do
+              let delayMicros = natToNum @StartDelay * 1_250_000
+              threadDelay delayMicros
+              putStrLn [i|Slept for: #{delayMicros}μs|]
+              newCurrentTime <- muGetCurrentTime (L.head targets) (L.head muGdbs)
+              putStrLn [i|Clock is now: #{newCurrentTime}|]
+
+            _ <- liftIO $ sequenceA $ L.zipWith muReadPeBuffer targets muGdbs
+
+            let
+              finalExit =
+                fromMaybe ExitSuccess
+                  $ L.find (/= ExitSuccess) [stabilityExitCode, gdbExitCode0, gdbExitCode1]
+            return finalExit
+
+printAllIgns ::
+  forall n m l.
+  ( HasCallStack
+  , KnownNat m
+  , KnownNat n
+  , KnownNat l
+  , l + 1 ~ (n + 1) * m
+  ) =>
+  Vec (n + 1) (Vec m (Unsigned 64, Unsigned 64)) ->
+  Vec (n + 1) (FpgaId, Vec m (Index (n + 1))) ->
```

Comment:
I think UGN/IGN calculation stuff should be moved to `Calculator`. Let's have one place to deal with all these tables.